Libpitt/member info 4

libpitt opened the pull request:
Please have a look at this @yuanzhou. 

`Download Member Info` link added under the admin scope in the `main_layout.html` 
Methods added to search db and construct csv for download.

Upon click, the UI shows a 3 second toast to indicate the request was received and a spinner near the link. When the response from the server is returned, the spinner goes back to hidden state.
<img width="952" alt="member-downloads" src="https://user-images.githubusercontent.com/115558393/200037947-4286218a-7b89-4b63-adb3-b47dda01507d.png">

Check it out. Let me know if I missed anything.

cc @shirey

libpitt committed:
Add gitingore

libpitt committed:
Add ability to download as csv (not done) - #4

libpitt committed:
Add functionality to download csv data - #4

libpitt committed:
Add toastify and custom style and js - #4

libpitt committed:
Remove test code

libpitt committed:
Add link and static files to layout - #4

libpitt committed:
Update gitignore

libpitt committed:
Uncomment admin_required decorator - #4

libpitt committed:
Reformat indents

libpitt committed:
Add blank line

libpitt committed:
Update gitingore

libpitt committed:
Neaten code

shirey reviewed (approved):
This generally looks good to me, but I'm the wrong person to review the code here.  I defer to @yuanzhou.  Once it has been deployed to DEV let me know and I'll check it out by using it.  Thanks, -Bill

libpitt committed:
Update casing

yuanzhou commented on static/js/main.js:
```diff
@@ -0,0 +1,127 @@
+
```
Can we rename this js file to `members-download.js`? Since it's only used for the download stuff. `main.js` is normally for the shared code.

libpitt replied:
@yuanzhou  The idea is that currently we have a bunch of js directly embedded into various html files. I was thinking in the future we have a bunch of separate modules that when compiled will just be a main.js file. But for now, I guess I can rename it because that functionality is not there.

yuanzhou commented on templates/layouts/main_layout.html:
```diff
@@ -1,15 +1,17 @@
+    <script type="text/javascript" src="{{ url_for('static', filename='js/fontawesome-all.min.js') }}"></script>
+    <script type="text/javascript" src="{{ url_for('static', filename='js/jquery.min.js') }}"></script>
+    <script type="text/javascript" src="{{ url_for('static', filename='bootstrap/js/bootstrap.min.js') }}"></script>
+    <script type="text/javascript" src="{{ url_for('static', filename='js/toastify-js.js') }}"></script>
```
We should only load the `toastify.min.css`, `toastify-js.js`, and the `members-download.js` (`main.js` before renaming) for Admin users. Just put them within the `{% if session['isAdmin'] %}  load the resources here  {% endif %}` block should be good. Won't hurt if loading them all the time, consider this as a performance tuning based on the actual needs.

libpitt replied:
@yuanzhou  Various other places throughout the app might need these features, so same as above reason.

yuanzhou commented on templates/layouts/main_layout.html:
```diff
@@ -21,19 +23,25 @@
+        </li>
+        <li class="nav-item">
+            <a class="nav-link state-element" id="js-btn--downloads-members" href="#" data-js-apicall="{{ config['FLASK_APP_BASE_URI'] }}/downloads/members" data-js-toastit="Your download will begin shortly ..." data-filename="SenNet-members-list.csv">
+                Download Member Info
```
Can we move this download link to the Members page? 

![image](https://user-images.githubusercontent.com/195873/200908932-0558251c-5ff7-4f75-a365-336e2ce54b6a.png)

libpitt replied:
@yuanzhou  Sure, I put the link there because this the original request from the the issue #4 card. Can you confirm you still want it moved?

yuanzhou replied:
Bill wanted to keep it on the Nav menu.

yuanzhou commented on app.py:
```diff
@@ -1998,6 +1998,82 @@ def unlink_globus_identities():
+    wp_users = WPUser.query.order_by(WPUser.id.desc()).all()
+    for user in wp_users:
+        # Check if this target user is a member (capabilities will be empty dict if not member role)
+        capabilities = next((meta for meta in user.metas if (meta.meta_key == wp_db_table_prefix + 'capabilities') and ('member' in meta.meta_value)), {})
```
I think we should also include the `administrator` to the download, not only `member` role. 

Let's also remove the "(not including the Administrators)" wording. And you'll also need to add the `administrator` role to https://github.com/sennetconsortium/member-ui/blob/libpitt/member-info-4/app.py#L1224 for consistency.

![image](https://user-images.githubusercontent.com/195873/200910599-386e80bf-7a64-4967-8426-6a93a3836857.png)

yuanzhou commented on app.py:
```diff
@@ -1998,6 +1998,82 @@ def unlink_globus_identities():
+        'username': session['name']
+    }
+    members = get_all_members_with_all_info()
+    cvs = 'Globus Username Associated, Name, Email, Component, PM, PM email, Role, SenNet Data Via Globus, Gdrive account, GitHub, Slack, protocols.io  \n'
```
We should have both `globus_user_id` and `globus_username` in the CSV columns.

libpitt replied:
@yuanzhou Columns included according to the sample csv file that was provided in issue https://github.com/sennetconsortium/member-ui/issues/4 . Please confirm you still want this changed.

yuanzhou commented on app.py:
```diff
@@ -1998,6 +1998,82 @@ def unlink_globus_identities():
+                        else:
+                            obj = next((meta for meta in connection_data.metas if meta.meta_key == connection_meta_key_prefix + k), {'meta_value': ''})
+                            if obj is not None:
+                                member[k] = obj['meta_value'] if isinstance(obj, (dict)) else obj.meta_value
```
Here we should overwrite the value of `component`, `role`, `organization` with the actual value that the user entered for `other_component`, `other_role`, `other_organization`. Otherwise we'll only see "Other" in those columns instead of the actual user provided value due to the handling at https://github.com/sennetconsortium/member-ui/blob/libpitt/member-info-4/app.py#L2065-L2070

libpitt replied:
@yuanzhou Don't quite understand this request. Will ping you.

yuanzhou commented on app.py:
```diff
@@ -1998,6 +1998,82 @@ def unlink_globus_identities():
+
+    for member in members:
+        i = StageUser(member)
+        cvs += f"{i.globus_username},\"First name: {i.first_name}\nLast name:{i.last_name}\",{i.email},"
```
Let's just merge the first name and last name without using "First name: " and "Last name: " indicators. And using the `\n` would result two lines of the value.

libpitt replied:
@yuanzhou This is formatted according to the sample csv file that was provided in issue #4 . Please confirm you still want this changed.

yuanzhou reviewed (changes requested):
@libpitt please let me know if you have any questions. I've made sure your role is Admin in the DEV database.

libpitt committed:
Add PR changes:
First name and last name to two separate columns
Add user details for component, role and organization as Other:
Add access_requests column and data
Admins included in list

libpitt committed:
Use common JS library

yuanzhou commented on app.py:
```diff
@@ -1998,26 +1999,30 @@ def unlink_globus_identities():
                 # Also get the globus_user_id
                 wp_user_meta_globus_user_id = WPUserMeta.query.filter(WPUserMeta.user_id == user.id, WPUserMeta.meta_key.like('openid-connect-generic-subject-identity')).first()
+                # Get the user Capability
+                capability = 'administrator' if 'admin' in capabilities.meta_value else 'member'
```
@libpitt let's use `administrator` instead of `admin` for consistency.

yuanzhou commented on app.py:
```diff
@@ -2028,7 +2033,8 @@ def get_all_members_with_all_info():
                     'component': connection_data.department,
                     'role': connection_data.title,
                     'bio': connection_data.bio,
```
We should remove the `Bio` field, we won't be showing that in the CSV

libpitt committed:
Remove bio, use full adminstrator wording

libpitt committed:
Use modern js practices

libpitt committed:
Update static-src/README

libpitt commented:
@yuanzhou 
This has been updated with your change requests. I additionally used modern JS to facilitate scaling in the future instead of the the very verbose vanilla JS written earlier. 

Thanks!

libpitt committed:
Rename for consistency and semantics

libpitt committed:
Add file name changes

libpitt committed:
Complete change requests:
Change 'SenNet Data Via Globus' as the header for globus_identity to 'Globus Identity'
Move globus_identity column to the second column (after globus_username)
Fix formatting of access_requests
Update filename for blob response

libpitt commented on app.py:
```diff
@@ -2071,15 +2071,21 @@ def format_user_entry(val, other):
+        try:
+            access_requests_array = ast.literal_eval(json.loads(i.access_requests))
+            access_requests_array.sort()
+            access_requests = "; ".join(access_requests_array)
```
@yuanzhou If you prefer, this can be joined for separate lines.

libpitt committed:
Enclose entire column field with double quotes to escape user values with commas
